Migration to Rocky8 spack-stack installations on Jet

InnocentSouopgui-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
<!-- This description will become the commit message for the PR-->
<!-- Please use this format for your description:

  Describe your changes. Focus on the *what* and *why*. The *how* will be evident from the changes. In particular, be sure to note any interface changes, such as command line syntax, that will need to be communicated to users.

  At the end of your description, please be sure to add the issue this PR solves using the word "Resolves". If there are any issues that are related but not yet resolved (including in other repos), you may use "Refs".

  Resolves #1234
  Refs #4321
  Refs NOAA-EMC/repo#5678
-->
This migrates Global Workflow to Rocky8 spack-stack installations on Jet.
Jet has moved from CentOS7 to Rocky8.

Resolves #2377
Refs NOAA-EMC/UPP#919
Refs NOAA-EMC/gfs-utils#60
Refs NOAA-EMC/GSI#732
Refs NOAA-EMC/GSI-Monitor#130
Refs NOAA-EMC/GSI-utils#33

# Type of change
<!-- Delete all except one -->
- Bug fix (fixes something broken)
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Cycled experiments (48+ hours) at resolutions
- [x] 96/48 on xjet and kjet
- [x] 192/96 on kjet
- [x] 384/192 on kjet

Forecast only experiment (48+ hours) at resolutions
- [x] 48
- [x] 96
- [x] 192
- [x] 384
<!--
Example:
- Clone and build on Jet
- Forecast-only on Hera
-->

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have made corresponding changes to the documentation if necessary
<!-- - [ ] New and existing tests pass with my changes -->

InnocentSouopgui-NOAA committed:
Adding perl as module to be loaded; On Jet/Rocky8 Perl packages will not be provided by the OS. This solves a crash in vminmon tasks

InnocentSouopgui-NOAA committed:
Update modulefiles/module_gwsetup.jet.lua
to use rocky8 environment

InnocentSouopgui-NOAA committed:
update versions/run.jet.ver to use rocky8 for spack_envwq

InnocentSouopgui-NOAA commented:
Successful cycling over 3 days at resolution 96/48 on xjet and kjet.
Higher resolution in progress.

InnocentSouopgui-NOAA commented:
@DavidHuber-NOAA
There is a need to update the config files on Jet. I need some guidances here to proceed.
There are two reasons for the need to update the config files:
1. The required memory per node for some jobs is beyond the available memory per node, so those jobs never start; for example:
    - esfc jobs requires 80GB of memory per node, xjet available memory per node is 61GB
    - upp jobs requires 96GB of memory per node, kjet available memory per node is 93GB
When the memory requirement is lowered, those jobs run.

2. The wall clock limit is too short for some jobs at high resolution; for example:
    - enkfgdasesfc: 8 min too short at 192/60 (bump to 15?)
    - gdasfcst: 15 min too short at 192/60 (bump to 20?)
    - sfcanl: 10 min too short at 384/192 (bump to 15?)

The required memory per node is the critical one. Should I add specific instruction like
```
if [[ ${machine} == "JET" && ${partition} == "kjet" ]]; then export memory_upp="80GB" ; fi
if [[ ${machine} == "JET" && ${partition} == "xjet" ]]; then export memory_upp="60GB" ; fi
```
in the config.resources?

For the wall clock, 
```
if [[ ${machine} == "JET" ]]; then export wtime_sfcanl="00:15:00" ; fi # for sfcanl jobs
```
What do you think?

DavidHuber-NOAA commented:
>     2. The wall clock limit is too short for some jobs at high resolution; for example:
>        
>        * enkfgdasesfc: 8 min too short at 192/60 (bump to 15?)
>        * gdasfcst: 15 min too short at 192/60 (bump to 20?)
>        * sfcanl: 10 min too short at 384/192 (bump to 15?)
These all sound fine to me.  I think you can make these global (not machine-specific), since they are relatively small bumps.  Also, we don't typically test at C192, so these updates may be needed across the board.

> The required memory per node is the critical one. Should I add specific instruction like
> 
> ```
> if [[ ${machine} == "JET" && ${partition} == "kjet" ]]; then export memory_upp="80GB" ; fi
> if [[ ${machine} == "JET" && ${partition} == "xjet" ]]; then export memory_upp="60GB" ; fi
> ```
How about just telling srun to use all available memory for Jet nodes:
```
if [[ ${machine} == "JET" ]]; then export memory_upp="0" ; fi
```

InnocentSouopgui-NOAA commented:
That sounds good, I will do that.

InnocentSouopgui-NOAA committed:
use all available memory on Jet for esfc and upp; kjet has 93GB avail/per node and xjet has 61GB available
Bump esfc time to 15 minutes, 8 min doesnt work on Jet for res 192 plus.
bumb gdasfcst wall clock to 20 minutes for resolution 192 an bellow, 15 min not enough for 192 on Jet
bump sfcanl wall clock to 15min

InnocentSouopgui-NOAA commented:
gfsfcst task is hanging in forecast-only experiment. I have the end of the log file below in case it rings a bell.
That is about 3 minutes of wall-clock time into the run and it does not not move past that point.
I am running a test at the resolution 48 on Jet.

@aerorahul do you have any idea?

==== section of the log file ===
```
33:  EXPTOTAL is selected
33:  But PDLIB is not
36:  EXPTOTAL is selected
36:  But PDLIB is not
38:  EXPTOTAL is selected
38:  But PDLIB is not
39:  EXPTOTAL is selected
39:  But PDLIB is not
24:             Fields   : no fields defined
24:  
24:   Time interval :  
24:  --------------------------------------------------
24:        Starting time : 2021/03/23 12:00:00 UTC
24:        Ending time   : 2021/03/28 12:00:00 UTC
24:  
24:  
24:   Output requests :  
24:  --------------------------------------------------
24:        No dedicated output process, any file system.
24:  
24:        Type 1 : Fields of mean wave parameters
24:       -----------------------------------------
24:             From     : 2021/03/23 12:00:00 UTC
24:             To       : 2021/03/28 12:00:00 UTC
24:             Interval :            01:00:00
24:  
24:  
24:        Type 2 : Point output                   
24:       -----------------------------------------
24:             From     : 2021/03/23 12:00:00 UTC
24:             To       : 2021/03/28 12:00:00 UTC
24:             Interval :            01:00:00
24:  
24:  
24:        Type 8 : Restart files second request   
24:       -----------------------------------------
24:             From     : 2021/03/23 12:00:00 UTC
24:             To       : 2021/03/28 12:00:00 UTC
24:             Interval :            12:00:0
24:  
24:        Wave model ...
24:  EXPTOTAL is selected
24:  But PDLIB is not
```

JessicaMeixner-NOAA commented:
@InnocentSouopgui-NOAA this is a  compile time vs runtime mismatch issue in the wave model.  What test are you running when you get this error?

InnocentSouopgui-NOAA commented:
> @InnocentSouopgui-NOAA this is a compile time vs runtime mismatch issue in the wave model. What test are you running when you get this error?
I am running a forecast-only S2SW experiment using the prototype_ICs.

JessicaMeixner-NOAA commented:
Prototype ICs, likely mean you are trying to run an unstructured grid for the waves, which means you need to add -w when compiling the model.

aerorahul commented:
@InnocentSouopgui-NOAA 
When building and running for `system = gfs`  cases, do not use the `-w` option with `build_all.sh`.
The `-w` option should only be used with the `system = gefs` cases where a structured WW3 grid is used.

InnocentSouopgui-NOAA commented:
@JessicaMeixner-NOAA I recompiled gfs with -w and it is running now without hanging.
However with the comment from @aerorahul about not compiling gfs with -w, Ihave more questions:
The goal here is to test forecast only on Jet/Rocky8 before merging everything and have it available for other people who run on Jet.

What is the best course of action?

DavidHuber-NOAA commented:
@InnocentSouopgui-NOAA The meaning of the `-w` flag for `build_all.sh` changed recently in the develop branch.  It used to mean 'use the unstructured wave grid', but now it means 'use the structured wave grid'.  If you sync your branch with develop, you will capture this update.  Once doing so, you _**should not**_ use `-w` when running GFS experiments.

InnocentSouopgui-NOAA commented:
> @InnocentSouopgui-NOAA The meaning of the `-w` flag for `build_all.sh` changed recently in the develop branch. It used to mean 'use the unstructured wave grid', but now it means 'use the structured wave grid'. If you sync your branch with develop, you will capture this update. Once doing so, you _**should not**_ use `-w` when running GFS experiments.

Great, thanks. I remember seeing a notification about the change in the meaning of -w flag.

InnocentSouopgui-NOAA commented:
@DavidHuber-NOAA, Pretty much everything is ready to submit this pull request. I am waiting for the last forecast only experiment to complete.
Question: who maintains the external dependencies on Jet?
TC_tracker, OBSPROC and Fit2Obs were compiled with with stack spack 1.6.0 and link from my personal directory for all tests.
Fit2Obs source code was Downloaded from https://github.com/NOAA-EMC/Fit2Obs/archive/refs/tags/wflow.1.1.0.tar.gz. TC_tracker and OBSPROC source code was copied from the official BASE_GIT on Jet and recompiled.

DavidHuber-NOAA commented:
@InnocentSouopgui-NOAA Those are maintained by @KateFriedman-NOAA on Jet.  Kate, would you be able to get Innocent's branches installed into the official locations?

Innocent, when that's finished, you will want to update the locations in your branch.  I believe they will end up in /lfs4/HFIP/hfv3gfs/glopara/git.

KateFriedman-NOAA commented:
@InnocentSouopgui-NOAA Can you point me to your branches for each of these repos? It's a bit of a mixed bag for each external piece:

Fit2Obs:
- latest install on Jet: /lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.1.0
- Tag on GitHub: https://github.com/NOAA-EMC/Fit2Obs/releases/tag/wflow.1.1.0
- We need to roll any changes you made into the Fit2Obs `develop` branch and make a new tag (which I would then install everywhere)
- An issue needs to be opened in Fit2Obs as well (to document the Jet update): https://github.com/NOAA-EMC/Fit2Obs/issues

OBSPROC (actually it's the prepobs repo, we don't use any execs from the obsproc repo so I only update that when there is a science/script change, OBSPROC is the umbrella name for the obsproc, prepobs, other related repos that we don't need; we just use obsproc and prepobs):
- need to get Jet Rocky8 updates into our branch with prepobs via PR: https://github.com/NOAA-EMC/prepobs/tree/dev/gfsv17
- then I can install the `dev/gfsv17` branch on Jet (we haven't installed it there yet since the prior install still worked before)

TC_tracker:
- similar to prepobs, we are operating out of a branch currently: https://github.com/NOAA-EMC/TC_tracker/tree/feature/GFSv17_com_reorg
- need to get updates into that branch via PR

Once the above tags/branches are updated I can install them in the glopara space on Jet. Need to test with personal branches until then. Let me know how I can assist with getting updates into those repos.

KateFriedman-NOAA commented:
I should also say...you can make me the reviewer and/or assignee on any PRs for those three repos. I can review, merge, and tag in those repos.

InnocentSouopgui-NOAA commented:
@KateFriedman-NOAA
I will open the issue for those external dependencies and bring the changes in the repos. Since repositories were so outdated, still using stack-spack, I copied the versions already in /lfs4/HFIP/hfv3gfs/glopara/git and recompiled with spack-stack. And downloaded the tarball of Fit2Obs that matched what was in /lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.1.0.

aerorahul commented:
> @KateFriedman-NOAA I will open the issue for those external dependencies and bring the changes in the repos. Since repositories were so outdated, still using stack-spack, I copied the versions already in /lfs4/HFIP/hfv3gfs/glopara/git and recompiled with spack-stack. And downloaded the tarball of Fit2Obs that matched what was in /lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.1.0.

@InnocentSouopgui-NOAA 
It is on our TODO to make obsproc, prepobs, fit2obs, and tracker repos as submodules.

KateFriedman-NOAA commented:
> I will open the issue for those external dependencies and bring the changes in the repos.

Sounds good, thanks @InnocentSouopgui-NOAA ! Let me know if you need me at all ahead of opening PRs.

InnocentSouopgui-NOAA commented:
@KateFriedman-NOAA for prepobs, it looks like there other steps in addition to build to get everything in  place.

Let's break it down a little more.

I cloned the prepobs repository (dev/gfsv17 branch), built and used it to replace the opbproc, and the prep tasks are failing because of missing scripts. I looked through the logs and see that there are files missing from the repository.

Before the migration to rocky8, /mnt/lfs4/HFIP/hfv3gfs/glopara/git/obsproc/v1.1.2 is used in global workflow.
If I understood your explanation above, that obsproc is the prepobs repo then, I am missing few steps to get everything in place. For instance, those sub-directories are missing:
 - jobs: a set of scripts used by some jobs
 - parm: set of parameters files or so.
 - exec : set of executables of so.
 - docs
 - ecf
 - triggers

Thoughts?

KateFriedman-NOAA commented:
@InnocentSouopgui-NOAA Apologies for any confusion. Let me be more clear about obsproc/prepobs:

We use both obsproc and prepobs repos (under the uppercase OBSPROC umbrella of repos) but we only use execs from prepobs so we've only had to update prepobs for ports/libraries. There will be an obsproc update in ops soon (to v1.2) but I'll handle that install when it is ready.

You can leave the obsproc setting in `config.base` as is. You'll need to change the prepobs module path in `modulefiles/module_base.jet.lua`. Here: https://github.com/NOAA-EMC/global-workflow/blob/develop/modulefiles/module_base.jet.lua#L45

Point to your test copy there and then when I install the official copy under the glopara space we'll need to change the `feature-GFSv17_com_reorg_log_update` to the new install name. It won't have a version # yet until our changes make it into the version that goes into ops (down the road) so we'll have a non-versioned name for now.

Let me know if you have any questions. :)

InnocentSouopgui-NOAA commented:
@KateFriedman-NOAA ,

Thank you for the clarifications. let's give it another iteration.
Now that you pointed me to `modulefiles/module_base.jet.lua` I did not account for that one. It is working just fine as it is. Possibly the executables were linked statically. Since it is working just fine, we can keep it as is for the time being.

For now, only the tasks taking the settings from config.* are using the recompiled external modules, namely:
 - obsproc setting in config.base
 - TC_tracker setting in config.tropcy

The the following two are working just fine without being recompiled:
- Fit2Obs (/lfs4/HFIP/hfv3gfs/glopara/git/Fit2Obs/v1.0.0) the version used is 1.0.0
- prepobs (/lfs4/HFIP/hfv3gfs/glopara/git/prepobs/feature-GFSv17_com_reorg_log_update)

I am tempted to test with the existing TC_tracker and obsproc in /lfs4/HFIP/hfv3gfs/glopara/git ; If it works just fine as Fit2Obs and prepobs, we can keep all those external dependencies as is for now, and migrate them later. I can keep the draft pull requests in place.

Your thoughts!

KateFriedman-NOAA commented:
> I am tempted to test with the existing TC_tracker and obsproc in /lfs4/HFIP/hfv3gfs/glopara/git ; If it works just fine as Fit2Obs and prepobs, we can keep all those external dependencies as is for now, and migrate them later.

@InnocentSouopgui-NOAA If the rest of the system (particularly the critical components like GSI, UWM, UPP, UFS_UTILS, etc.) work in your branch and the current installed versions of those externals still work post-Rocky8-transition then I agree...let's finish them up later and get the needed Rocky8 updates in for the critical parts of the system.

Let me know if the current TC_tracker and obsproc work in your testing. Thanks!

InnocentSouopgui-NOAA commented:
> Let me know if the current TC_tracker and obsproc work in your testing. Thanks!
@KateFriedman-NOAA 
- obsproc works just fine on one cycle
- TC_tracker is broken because of netcdf libraries and a couple of problems in other scripts.

I am doing few more tests and the I will all the pull request, but the only one that will need to be re-installed before the global workflow pull request is TC_tracker

InnocentSouopgui-NOAA committed:
Merge branch 'NOAA-EMC:develop' into migration/jet-rocky8

InnocentSouopgui-NOAA commented:
@KateFriedman-NOAA,
Are you done with the version upgrade of fit2obs everywhere?
That way, I can add it to this pull request.

KateFriedman-NOAA commented:
> @KateFriedman-NOAA, Are you done with the version upgrade of fit2obs everywhere? That way, I can add it to this pull request.

Yes, please proceed. @WalterKolczynski-NOAA is waiting for Orion to be back and then will install the new Fit2Obs tag there. Don't wait for that now, please proceed.

DavidHuber-NOAA reviewed (commented):
Do any submodules need to be updated for this?

InnocentSouopgui-NOAA commented:
> Do any submodules need to be updated for this?

All the submodules have been updated and the pull requests merged.
As of now there is only one external dependency (TC_tracker) that is still waiting for review.

aerorahul commented:
@InnocentSouopgui-NOAA Can you please resolve the conflicts?

aerorahul commented on parm/config/gfs/config.resources:
```diff
@@ -734,6 +734,7 @@ case ${step} in
         exit 4
       ;;
     esac
+    if [[ ${machine} == "JET" ]]; then export memory_upp="0" ; fi
```
What happens if it get "some" memory?

InnocentSouopgui-NOAA replied:
As in the case of esfc, xjet and kjet have very different size of memory per nodes. Setting 0 (to use all available memory) made sense to keep the modifications to minimal and allow higher resolution to run on kjet while lower res can run on xjet.

aerorahul replied:
would `unset memory_upp` work instead of `export memory_upp=0`?

InnocentSouopgui-NOAA replied:
I don't know if `unset memory_upp` would work or not. I have not used that option before. Let me tag @DavidHuber-NOAA to see if he tried that before.

InnocentSouopgui-NOAA replied:
@aerorahul I changed to `unset memory_upp` for jet

aerorahul commented on parm/config/gfs/config.resources:
```diff
@@ -1013,13 +1014,14 @@ case ${step} in
     export nth_cycle=${nth_esfc}
     export npe_node_cycle=$(( npe_node_max / nth_cycle ))
     export memory_esfc="80GB"
+    if [[ ${machine} == "JET" ]]; then export memory_esfc="0" ; fi
```
What happens if it get "some" memory?
Would it be better to cast it as:
```
if [[ ${machine} != "JET" ]]; then export memory_esfc="80GB" ; fi
```

InnocentSouopgui-NOAA replied:
> What happens if it get "some" memory? Would it be better to cast it as:
> 
> ```
> if [[ ${machine} != "JET" ]]; then export memory_esfc="80GB" ; fi
> ```

That will not work on xjet partition; Xjet has a max of 64G/node.

WalterKolczynski-NOAA replied:
That would specifically be if it wasn't on Jet. On Jet the value would be unset.

InnocentSouopgui-NOAA replied:
My bad, I read incorrectly. I get it now.
I will set a run with that and see how it goes.
Same with memory_upp, I will test both together.

InnocentSouopgui-NOAA replied:
@aerorahul, I changed to your suggestion, setting `memory_esfc` only if not on JET.

InnocentSouopgui-NOAA committed:
Merge pull request #2 from NOAA-EMC/develop

Merging the base develop branch into this one

InnocentSouopgui-NOAA committed:
Resolve conflicts with the base repos develop branch

InnocentSouopgui-NOAA commented:
> @InnocentSouopgui-NOAA Can you please resolve the conflicts?

I just resolved the conflicts. I am now testing unsetting variables on JET instead on setting it to zero.

InnocentSouopgui-NOAA committed:
Unset memory for JET instead of setting to 0

InnocentSouopgui-NOAA committed:
Merge pull request #3 from NOAA-EMC/develop

Mirging additional changes from base

aerorahul reviewed (approved):
looks good.

WalterKolczynski-NOAA commented:
Don't the submodules in workflow need to be updated to the new hashes that include your changes to those repos?

InnocentSouopgui-NOAA commented:
> Don't the submodules in workflow need to be updated to the new hashes that include your changes to those repos?

Sure they need to be updated, I am not yet familiar with that part.
What do I need to do?

DavidHuber-NOAA commented:
For each of the submodules, you need to locally check out the versions that you need, then `git add; git commit` them.  For instance, with the GSI, it looks like you updated the Jet modules at https://github.com/NOAA-EMC/GSI/commit/8e279f9c734097f673b07e80f385b2623d13ba4a, so you would:
```bash
cd <global-workflow>/sorc/gsi.fd
git fetch
git checkout 8e279f9c
cd ../
git add gsi.fd
git commit -m "Update gsi hash to include Jet's Rocky8 upgrade"
```

KateFriedman-NOAA commented:
@InnocentSouopgui-NOAA In addition to updating the component submodules (thanks @DavidHuber-NOAA for providing those instructions!), the `fit2obs_ver` in `versions/run.spack.ver` needs to be updated to `1.1.0`. See this line to make the change:
https://github.com/NOAA-EMC/global-workflow/blob/develop/versions/run.spack.ver#L35

Are we still waiting on the TC_tracker updates to go in? If so, we can take care of that afterwards, let's get the submodules and such wrapped up and merge this PR. Thanks!

InnocentSouopgui-NOAA committed:
update the hash for submodules that are still outdated

InnocentSouopgui-NOAA committed:
Update fit2obs version

InnocentSouopgui-NOAA commented:
@KateFriedman-NOAA,
What about  `fit2obs` version `1.1.1` you mentioned last week or so? Is it not ready yet?

KateFriedman-NOAA commented:
> @KateFriedman-NOAA, What about `fit2obs` version `1.1.1` you mentioned last week or so? Is it not ready yet?

It's ready, sorry I mentioned `1.1.0` and should have said `1.1.1`. The `1.1.1` is installed. You can proceed with updating `fit2obs_ver=1.1.1`.

InnocentSouopgui-NOAA committed:
Update fit2obs version

InnocentSouopgui-NOAA commented:
Submodules updated, fit2obs version updated as well.

KateFriedman-NOAA reviewed (approved):
Updates for Rocky8 on Jet look good. Thanks for your work on this @InnocentSouopgui-NOAA !

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of CI tests.

DavidHuber-NOAA reviewed (approved):
Changes look good.  Thanks!
